reactioncommerce/reaction
# Add Order.referenceId

Resolves #4766  
Impact: **minor**  
Type: **feature**

Related storefront PR: https://github.com/reactioncommerce/reaction-next-starterkit/pull/434

## Issue
Orders did not have a non-transformed reference ID on them, by which they could be referenced and matched up by both operators and customers. This was being worked around in storefront code by decoding the `Order._id` from GraphQL, but this is not a reliable way to do it, and it doesn't allow for customization of order reference ID generation.

## Solution
- Add `Order.referenceId` both in the database and in GraphQL schema
- Migrate existing orders to set the internal `Order._id` to the `Order.referenceId`
- Update all operator and classic storefront UI that shows "Order ID" to show the `Order.referenceId` instead of the `Order._id`.

## Breaking changes
None

## Testing
1. Place an order using the classic UI
2. Verify that the order ID on the order completion page matches the order ID that shows everywhere in the operator Orders panel, on PDF invoice, and in all emails that are sent.
3. Place an order using the NextJS storefront UI (branch from corresponding PR https://github.com/reactioncommerce/reaction-next-starterkit/pull/434)
4. Verify that the order ID on the order completion page matches the order ID that shows everywhere in the operator Orders panel, on PDF invoice, and in all emails that are sent.
5. Query the orders by GraphQL and verify that you can get the `order.referenceId` field.
